### PR TITLE
Improve credential documentation and expired token error raising

### PR DESCRIPTION
**Pull request by mullermp**
Improve the documentation for credential providers.

With an expired token, the ExpiredToken error is raised instead of a cryptic OpenSSL socket error.

### CODE REVIEW

File: gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb [hunk 1]
```diff
@@ -3,25 +3,28 @@
...
   #     )
-  #
   #     ec2 = Aws::EC2::Client.new(credentials: role_credentials)
   #
-  # If you omit `:client` option, a new {STS::Client} object will be
-  # constructed.
+  # If you omit `:client` option, a new {Aws::STS::Client} object will be
+  # constructed with additional options that were provided.
+  #
+  # Automatically handles refreshing credentials if an Expiration time is
```
Comment (alextwoods) [inline gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb]: What does if an expiration time is provided in the payload mean?  This to me sounds like something a user needs to do to get this behavior.

File: gems/aws-sdk-core/lib/aws-sdk-core/assume_role_web_identity_credentials.rb [hunk 2]
```diff
@@ -16,12 +15,20 @@ module Aws
...
-  #     @see Aws::STS::Client#assume_role_with_web_identity
+  #     ec2 = Aws::EC2::Client.new(credentials: role_credentials)
+  #
+  # If you omit `:client` option, a new {Aws::STS::Client} object will be
+  # constructed with additional options that were provided.
+  #
+  # Automatically handles refreshing credentials if an Expiration time is
+  # provided in the credentials payload.
   #
+  # This credential provider also provides a `before_refresh` callback
```
Comment (alextwoods) [inline gems/aws-sdk-core/lib/aws-sdk-core/assume_role_web_identity_credentials.rb]: I'm not sure we need to document this here - its already documented on the client initialize options.

File: gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb [hunk 3]
```diff
@@ -3,25 +3,28 @@
 require 'set'
 
 module Aws
-
-  # An auto-refreshing credential provider that works by assuming
-  # a role via {Aws::STS::Client#assume_role}.
+  # A credential provider that assumes a role via
```
Comment (alextwoods) [inline gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb]: Nit: Why remove the auto-refreshing from all of these descriptions?

File: gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb (hunk 1 above)
Comment (mullermp) [reply gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb]: If AssumeRole returns Expiration in the JSON body, it's used here.

File: gems/aws-sdk-core/lib/aws-sdk-core/assume_role_web_identity_credentials.rb (hunk 2 above)
Comment (mullermp) [reply gems/aws-sdk-core/lib/aws-sdk-core/assume_role_web_identity_credentials.rb]: Assume Role Credentials had it already in the Class documentation (but that was the only instance). I can remove all of them, or keep all of them added here. Which do you prefer?

File: gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb (hunk 3 above)
Comment (mullermp) [reply gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb]: It was already mentioned multiple times (I made it generically below, too)

File: gems/aws-sdk-core/lib/aws-sdk-core/assume_role_web_identity_credentials.rb (hunk 2 above)
Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/assume_role_web_identity_credentials.rb]: I think it likely makes more sense to remove from class documentation and keep in initialize options.

File: gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb (hunk 1 above)
Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb]: How do I as a customer know if my credentials will be refreshing or if the expiration will be in the json body of the response?  
I'd lean towards leaving the "Auto refreshing" in the description.  Or for credentials like process credentials where the user's custom process has control over the behavior, then document it like this.

**Review by alextwoods — APPROVED** [review]

**Pull request merged**